reference-implementation/lib/readable-stream.js

```diff
@@ -392,12 +398,12 @@ function ReadableStreamPipeTo(source, dest, preventClose, preventAbort, preventC
     // - Backpressure must be enforced
     // - Shutdown must stop all activity
     function pipeLoop() {
-      return new Promise((resolveLoop, rejectLoop) => {
+      return newPromise((resolveLoop, rejectLoop) => {
         function next(done) {
           if (done) {
             resolveLoop();
           } else {
-            pipeStep().then(next, rejectLoop);
+            uponPromise(pipeStep(), next, rejectLoop);
```

Comment:
I'm not sure if `uponPromise()` makes this code more readable. I feel like `PerformPromiseThen(pipeStep(), next, rejectLoop)` is more obvious, but then we have to export the low-level `then()` helper again. 😕

This also adds an extra `.catch()` handler on every pipe step, since this is now equivalent to:
```javascript
pipeStep().then(next, rejectLoop).catch(rethrowAssertionErrorRejection);
```
The `.catch()` is not necessary, since we already handle that [here](https://github.com/whatwg/streams/blob/1146228f0bcc1077f6a0c0eff2452b89332ee2ba/reference-implementation/lib/readable-stream.js#L469):
```javascript
setPromiseIsHandledToTrue(pipeLoop());
```
which is equivalent to:
```javascript
pipeLoop().catch(rethrowAssertionErrorRejection);
```
Although you wouldn't want to create this extra promise in a real implementation, I think it's fine for a reference implementation.

Reply:
I think this use of `uponPromise()` is fine.